From 1e4413189bec283d89962d62b93100f7033a19ff Mon Sep 17 00:00:00 2001
From: Jun Zhang <jun.zhang@arm.com>
Date: Fri, 24 Nov 2023 21:03:16 +0800
Subject: [PATCH] rss/kronos: Handle host ack and fmp image info request

`fwu_host_ack` is used to handle the ACK from u-boot
which indicate whether the updated image can boot
successfully or not. The updated image will be
accepted if the ACK is reached before time out(360s),
rollback operation will try if the time out is
reached. The main logic is come from corstone1000
platform in `fw_agent/fwu_agent.c` file. Main changes:

1.Enable SysTick_IRQn timer in RSS runtime
2.Set security counter to OTP when updating images
3.`fmp_get_image_info` is used to get fmp image info
  for Trusted Service

Upstream-Status: Pending
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
---
 .../ext/target/arm/rss/common/CMakeLists.txt  |   6 +
 .../common/services/src/tfm_platform_system.c |  12 +
 .../ext/target/arm/rss/common/target_cfg.c    |   9 +
 .../arm/rss/common/tfm_hal_multi_core.c       |   4 +
 .../ext/target/arm/rss/kronos/CMakeLists.txt  |   1 +
 .../arm/rss/kronos/bl2/security_cnt_bl2.c     | 133 ++++++
 .../ext/target/arm/rss/kronos/config.cmake    |   1 +
 .../rss/kronos/fw_update_agent/fwu_agent.c    | 423 ++++++++++++++++++
 .../rss/kronos/fw_update_agent/fwu_agent.h    |  29 ++
 9 files changed, 618 insertions(+)
 create mode 100644 platform/ext/target/arm/rss/kronos/bl2/security_cnt_bl2.c

diff --git a/platform/ext/target/arm/rss/common/CMakeLists.txt b/platform/ext/target/arm/rss/common/CMakeLists.txt
index b2746b7ff..48ee219ec 100644
--- a/platform/ext/target/arm/rss/common/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/common/CMakeLists.txt
@@ -431,6 +431,12 @@ if(TFM_MULTI_CORE_TOPOLOGY)
         PRIVATE
             ${CMAKE_CURRENT_SOURCE_DIR}/tfm_hal_multi_core.c
     )
+
+    target_compile_definitions(tfm_psa_rot_partition_ns_agent_mailbox
+        PRIVATE
+            $<$<BOOL:${TFM_FWU_AGENT}>:TFM_FWU_AGENT>
+    )
+
 endif()

 #========================= tfm_spm ============================================#
diff --git a/platform/ext/target/arm/rss/common/services/src/tfm_platform_system.c b/platform/ext/target/arm/rss/common/services/src/tfm_platform_system.c
index 5f84a47ed..d3975a576 100644
--- a/platform/ext/target/arm/rss/common/services/src/tfm_platform_system.c
+++ b/platform/ext/target/arm/rss/common/services/src/tfm_platform_system.c
@@ -44,9 +44,21 @@ enum tfm_platform_err_t tfm_platform_hal_ioctl(tfm_platform_ioctl_req_t request,
         break;

     case IOCTL_FWU_HOST_ACK:
+        result = fwu_host_ack();
+        if (result != FWU_AGENT_SUCCESS) {
+            ret = TFM_PLATFORM_ERR_SYSTEM_ERROR;
+        }
         break;

     case IOCTL_FMP_GET_IMAGE_INFO:
+        if (out_vec == NULL) {
+            ret = TFM_PLATFORM_ERR_INVALID_PARAM;
+            break;
+        }
+        result = fmp_get_image_info(out_vec[0].base, out_vec[0].len);
+        if (result != FWU_AGENT_SUCCESS) {
+            ret = TFM_PLATFORM_ERR_SYSTEM_ERROR;
+        }
         break;

     default:
diff --git a/platform/ext/target/arm/rss/common/target_cfg.c b/platform/ext/target/arm/rss/common/target_cfg.c
index a1a45c814..35f0a4070 100644
--- a/platform/ext/target/arm/rss/common/target_cfg.c
+++ b/platform/ext/target/arm/rss/common/target_cfg.c
@@ -235,6 +235,10 @@ enum tfm_plat_err_t nvic_interrupt_target_state_cfg(void)
     NVIC_ClearTargetState(CMU_MHU8_Receiver_IRQn);
 #endif

+#ifdef TFM_FWU_AGENT
+    NVIC_ClearTargetState(SysTick_IRQn);
+#endif /* TFM_FWU_AGENT */
+
     return TFM_PLAT_ERR_SUCCESS;
 }

@@ -308,6 +312,11 @@ enum tfm_plat_err_t nvic_interrupt_enable(void)
     NVIC_EnableIRQ(CMU_MHU8_Receiver_IRQn);
 #endif

+#ifdef TFM_FWU_AGENT
+    NVIC_ClearPendingIRQ(SysTick_IRQn);
+    NVIC_EnableIRQ(SysTick_IRQn);
+#endif /* TFM_FWU_AGENT */
+
     return TFM_PLAT_ERR_SUCCESS;
 }

diff --git a/platform/ext/target/arm/rss/common/tfm_hal_multi_core.c b/platform/ext/target/arm/rss/common/tfm_hal_multi_core.c
index fcedcb111..334e04b6c 100644
--- a/platform/ext/target/arm/rss/common/tfm_hal_multi_core.c
+++ b/platform/ext/target/arm/rss/common/tfm_hal_multi_core.c
@@ -16,6 +16,10 @@ void tfm_hal_boot_ns_cpu(uintptr_t start_addr)

 void tfm_hal_wait_for_ns_cpu_ready(void)
 {
+#ifdef TFM_FWU_AGENT
+    /* Start the reset timer if firmware update process is ongoing */
+    host_acknowledgement_timer_to_reset();
+#endif /* TFM_FWU_AGENT */
     /* Nothing to do, all necessary synchronization done by BL2 */
     return;
 }
diff --git a/platform/ext/target/arm/rss/kronos/CMakeLists.txt b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
index bec00b70c..e99ae9b98 100644
--- a/platform/ext/target/arm/rss/kronos/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
@@ -107,6 +107,7 @@ target_sources(platform_ns
 target_sources(platform_bl2
     PRIVATE
         bl2/flash_map_bl2.c
+        bl2/security_cnt_bl2.c
         cmsis_drivers/Driver_Flash.c
         cmsis_drivers/Driver_USART.c
         device/host_device_definition.c
diff --git a/platform/ext/target/arm/rss/kronos/bl2/security_cnt_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/security_cnt_bl2.c
new file mode 100644
index 000000000..9fa281d25
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/bl2/security_cnt_bl2.c
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2019-2021,2023 Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+#include "bootutil/security_cnt.h"
+#include "tfm_plat_nv_counters.h"
+#include "tfm_plat_defs.h"
+#include "bootutil/fault_injection_hardening.h"
+#include <stdint.h>
+#include "tfm_plat_provisioning.h"
+#include "fwu_agent.h"
+#include "bootutil/security_cnt.h"
+#include "bootutil/fault_injection_hardening.h"
+#include "bootutil/bootutil.h"
+#include "bootutil/bootutil_log.h"
+#include <stdint.h>
+
+#define TFM_BOOT_NV_COUNTER_FIRST    PLAT_NV_COUNTER_BL2_0
+#define TFM_BOOT_NV_COUNTER_MAX      TFM_BOOT_NV_COUNTER_FIRST + \
+                                        MCUBOOT_IMAGE_NUMBER
+
+static enum tfm_nv_counter_t get_nv_counter_from_image_id(uint32_t image_id)
+{
+    uint32_t nv_counter;
+
+    /* Avoid integer overflow */
+    if ((UINT32_MAX - TFM_BOOT_NV_COUNTER_FIRST) < image_id) {
+        return TFM_BOOT_NV_COUNTER_MAX;
+    }
+
+    nv_counter = TFM_BOOT_NV_COUNTER_FIRST + image_id;
+
+    /* Check the existence of the enumerated counter value */
+    if (nv_counter >= TFM_BOOT_NV_COUNTER_MAX) {
+        return TFM_BOOT_NV_COUNTER_MAX;
+    }
+
+    return (enum tfm_nv_counter_t)nv_counter;
+}
+
+fih_ret boot_nv_security_counter_init(void)
+{
+    FIH_DECLARE(fih_rc, FIH_FAILURE);
+
+    fih_rc = fih_ret_encode_zero_equality(tfm_plat_init_nv_counter());
+
+    FIH_RET(fih_rc);
+}
+
+fih_ret boot_nv_security_counter_get(uint32_t image_id, fih_int *security_cnt)
+{
+    enum tfm_nv_counter_t nv_counter;
+    FIH_DECLARE(fih_rc, FIH_FAILURE);
+    uint32_t security_cnt_soft;
+
+    /* Check if it's a null-pointer. */
+    if (!security_cnt) {
+        FIH_RET(FIH_FAILURE);
+    }
+
+    nv_counter = get_nv_counter_from_image_id(image_id);
+    if (nv_counter >= TFM_BOOT_NV_COUNTER_MAX) {
+        FIH_RET(FIH_FAILURE);
+    }
+
+    fih_rc = fih_ret_encode_zero_equality(
+             tfm_plat_read_nv_counter(nv_counter,
+                                      sizeof(security_cnt_soft),
+                                      (uint8_t *)&security_cnt_soft));
+    *security_cnt = fih_int_encode(security_cnt_soft);
+
+    FIH_RET(fih_rc);
+}
+
+int32_t boot_nv_security_counter_update(uint32_t image_id,
+                                        uint32_t img_security_cnt)
+{
+    enum tfm_nv_counter_t nv_counter;
+    enum tfm_plat_err_t err;
+    enum fwu_agent_error_t fwu_err;
+    enum fwu_nv_counter_index_t tfm_nv_counter_i;
+
+    nv_counter = get_nv_counter_from_image_id(image_id);
+    if (nv_counter >= TFM_BOOT_NV_COUNTER_MAX) {
+        return -1;
+    }
+
+    if (tfm_plat_provisioning_is_required()) {
+        err = tfm_plat_set_nv_counter(nv_counter, img_security_cnt);
+        if (err != TFM_PLAT_ERR_SUCCESS) {
+            return -1;
+        }
+
+    } else {
+        switch (nv_counter) {
+        case PLAT_NV_COUNTER_BL2_0:
+            tfm_nv_counter_i = FWU_BL2_NV_COUNTER;
+            break;
+        case PLAT_NV_COUNTER_BL2_1:
+            tfm_nv_counter_i = FWU_TFM_NV_COUNTER;
+            break;
+        case PLAT_NV_COUNTER_BL2_2:
+            tfm_nv_counter_i = FWU_FIP_NV_COUNTER;
+            break;
+        case PLAT_NV_COUNTER_BL2_3:
+            tfm_nv_counter_i = FWU_SI_CL2_NV_COUNTER;
+            break;
+        case PLAT_NV_COUNTER_BL2_4:
+            tfm_nv_counter_i = FWU_SI_CL1_NV_COUNTER;
+            break;
+        case PLAT_NV_COUNTER_BL2_5:
+            tfm_nv_counter_i = FWU_SI_CL0_NV_COUNTER;
+            break;
+        case PLAT_NV_COUNTER_BL2_6:
+            tfm_nv_counter_i = FWU_LCP_NV_COUNTER;
+            break;
+        case PLAT_NV_COUNTER_BL2_7:
+            tfm_nv_counter_i = FWU_SCP_NV_COUNTER;
+            break;
+        default:
+            return -1;
+        }
+
+        fwu_err = fwu_stage_nv_counter(tfm_nv_counter_i, img_security_cnt);
+        if (fwu_err != FWU_AGENT_SUCCESS) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
diff --git a/platform/ext/target/arm/rss/kronos/config.cmake b/platform/ext/target/arm/rss/kronos/config.cmake
index df9f83959..a8090653b 100644
--- a/platform/ext/target/arm/rss/kronos/config.cmake
+++ b/platform/ext/target/arm/rss/kronos/config.cmake
@@ -37,6 +37,7 @@ set(TFM_PARTITION_CRYPTO                      ON   CACHE BOOL   "Enable Crypto p
 set(TFM_MBEDCRYPTO_PLATFORM_EXTRA_CONFIG_PATH  ""  CACHE PATH   "Config to append to standard Mbed Crypto config, used by platforms to cnfigure feature support")
 set(RSS_COMMS_MAX_CONCURRENT_REQ               4   CACHE STRING "Maximum number of concurrent connection requests to RSS")
 set(TFM_FWU_AGENT                              ON   CACHE BOOL   "Whether to enable FWU Agent or not")
+set(DEFAULT_MCUBOOT_SECURITY_COUNTERS          OFF  CACHE BOOL   "Whether to use the default security counter configuration defined by TF-M project")

 # Once all options are set, set common options as fallback
 include(${CMAKE_CURRENT_LIST_DIR}/../common/config.cmake)
diff --git a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.c b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.c
index a35c5f34a..b3f6c9b6f 100644
--- a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.c
+++ b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.c
@@ -27,6 +27,8 @@ extern ARM_DRIVER_FLASH FWU_METADATA_RSS_FLASH_DEV;
 extern int is_initialized_rss;
 extern struct fwu_metadata_rss fwu_md_rss;

+#define HOST_ACK_TIMEOUT_SEC    (6 * 60) /* About 360s, not exact. */
+
 capsule_image_info_t capsule_info;
 struct efi_guid full_capsule_image_guid = {
     .time_low = 0xe2bb9c06,
@@ -45,6 +47,41 @@ struct efi_guid full_capsule_image_guid = {
 int is_initialized_ap = 0;
 struct fwu_metadata_ap fwu_md_ap;

+/* Stage nv counter into private metadata section of the flash.
+ * Staged nv counters are written to the otp when firmware update is successful.
+ * The function assumes that the api is called in the boot loading stage.
+ */
+enum fwu_agent_error_t fwu_stage_nv_counter(enum fwu_nv_counter_index_t index,
+        uint32_t img_security_cnt)
+{
+    struct fwu_private_metadata priv_metadata;
+
+    FWU_LOG_MSG("%s: enter: index = %u, val = %u\n\r", __func__,
+                                index, img_security_cnt);
+
+    if (!is_initialized_rss) {
+        FWU_ASSERT(0);
+    }
+
+    if (index > FWU_MAX_NV_COUNTER_INDEX) {
+        return FWU_AGENT_ERROR;
+    }
+
+    if (private_metadata_read(&priv_metadata)) {
+        FWU_ASSERT(0);
+    }
+
+    if (priv_metadata.nv_counter[index] != img_security_cnt) {
+        priv_metadata.nv_counter[index] = img_security_cnt;
+        if (private_metadata_write(&priv_metadata)) {
+            FWU_ASSERT(0);
+        }
+    }
+
+    FWU_LOG_MSG("%s: exit\n\r", __func__);
+    return FWU_AGENT_SUCCESS;
+}
+
 uint8_t bl2_get_boot_bank(void)
 {
     uint8_t boot_index;
@@ -400,3 +437,389 @@ out:
     FWU_LOG_MSG("%s: exit: ret = %d\n\r", __func__, ret);
     return ret;
 }
+
+static enum fwu_agent_error_t accept_full_capsule(
+          struct fwu_metadata_rss* fwu_md_rss,
+          struct fwu_metadata_ap* fwu_md_ap,
+          struct fwu_private_metadata* priv_metadata)
+{
+    uint32_t active_index;
+    enum fwu_agent_error_t ret;
+
+    FWU_LOG_MSG("%s: enter\n\r", __func__);
+
+    if (!fwu_md_rss) {
+        return FWU_AGENT_ERROR;
+    }
+
+    active_index = fwu_md_rss->md.active_index;
+
+    for (uint8_t i = 0; i < NR_OF_IMAGES_IN_RSS_FW_BANK; i++) {
+        fwu_md_rss->img_entry[i].img_props[active_index].accepted =
+                                                            IMAGE_ACCEPTED;
+    }
+
+    active_index = fwu_md_ap->md.active_index;
+    for (uint8_t i = 0; i < NR_OF_IMAGES_IN_AP_FW_BANK; i++) {
+        fwu_md_ap->img_entry[i].img_props[active_index].accepted =
+                                                            IMAGE_ACCEPTED;
+    }
+
+    priv_metadata->boot_attempted = 0;
+
+    ret = private_metadata_write(priv_metadata);
+    if (ret) {
+        return ret;
+    }
+    fwu_md_rss->md.crc_32 = crc32((uint8_t *)&fwu_md_rss->md.version,
+                              sizeof(struct fwu_metadata_rss) - sizeof(uint32_t));
+
+    ret = metadata_write(FWU_RSS_FLASH_DEV, &fwu_md_rss->md);
+    if (ret) {
+        return ret;
+    }
+
+    fwu_md_ap->md.crc_32 = crc32((uint8_t *)&fwu_md_ap->md.version,
+                              sizeof(struct fwu_metadata_ap) - sizeof(uint32_t));
+
+    ret = metadata_write(FWU_AP_FLASH_DEV, &fwu_md_ap->md);
+    if (ret) {
+        return ret;
+    }
+
+    FWU_LOG_MSG("%s: exit: fwu state is changed to regular\n\r", __func__);
+    return FWU_AGENT_SUCCESS;
+}
+
+static enum fwu_agent_error_t fwu_accept_image(struct efi_guid* guid,
+                                               struct fwu_metadata_rss *fwu_md_rss,
+                                               struct fwu_metadata_ap *fwu_md_ap,
+                                               struct fwu_private_metadata *priv_metadata)
+{
+    enum fwu_agent_state_t current_state;
+    int image_index;
+    uint32_t image_bank_offset;
+    enum fwu_agent_error_t ret;
+
+    FWU_LOG_MSG("%s: enter\n\r", __func__);
+
+    /* It is expected to receive this call only when in trial state. */
+    current_state = get_fwu_agent_state(fwu_md_rss, priv_metadata);
+    if (current_state != FWU_AGENT_STATE_TRIAL) {
+        return FWU_AGENT_ERROR;
+    }
+
+    /* Booted from previous_active_bank, not expected
+     * to receive this call in this state, rather host should
+     * call fwu_select_previous.
+     */
+    if (fwu_md_rss->md.active_index != priv_metadata->boot_index) {
+        return FWU_AGENT_ERROR;
+    }
+
+    image_index = get_image_info_in_bank(guid, &image_bank_offset);
+    switch(image_index) {
+    case IMAGE_ALL:
+        ret = accept_full_capsule(fwu_md_rss, fwu_md_ap, priv_metadata);
+        break;
+    default:
+        FWU_LOG_MSG("%s: sent image not recognized\n\r", __func__);
+        ret = FWU_AGENT_ERROR;
+        break;
+    }
+
+    FWU_LOG_MSG("%s: exit: ret = %d\n\r", __func__, ret);
+    return ret;
+}
+
+static enum fwu_agent_error_t fwu_select_previous(
+                           struct fwu_metadata_rss *fwu_md_rss,
+                           struct fwu_metadata_ap *fwu_md_ap,
+                           struct fwu_private_metadata *priv_metadata)
+{
+    enum fwu_agent_error_t ret;
+    enum fwu_agent_state_t current_state;
+    uint32_t index;
+
+    FWU_LOG_MSG("%s: enter\n\r", __func__);
+
+    /* It is expected to receive this call only when in trial state. */
+    current_state = get_fwu_agent_state(fwu_md_rss, priv_metadata);
+    if (current_state != FWU_AGENT_STATE_TRIAL) {
+        return FWU_AGENT_ERROR;
+    }
+
+    /* Not expected to receive this call in this state, system
+     * did not boot from previous active index.
+     */
+    if (fwu_md_rss->md.previous_active_index != priv_metadata->boot_index) {
+        return FWU_AGENT_ERROR;
+    }
+
+    FWU_LOG_MSG("%s: trial state: active index = %u, previous active = %u\n\r",
+            __func__, fwu_md_rss->md.active_index, fwu_md_rss->md.previous_active_index);
+
+    index = fwu_md_rss->md.previous_active_index;
+    for (uint8_t i = 0; i < NR_OF_IMAGES_IN_RSS_FW_BANK; i++) {
+        if (fwu_md_rss->img_entry[i].img_props[index].accepted != IMAGE_ACCEPTED)
+        {
+            FWU_ASSERT(0);
+        }
+    }
+
+    for (uint8_t i = 0; i < NR_OF_IMAGES_IN_AP_FW_BANK; i++) {
+        if (fwu_md_ap->img_entry[i].img_props[index].accepted != IMAGE_ACCEPTED)
+        {
+            FWU_ASSERT(0);
+        }
+    }
+
+    index = fwu_md_rss->md.active_index;
+    fwu_md_rss->md.active_index = fwu_md_rss->md.previous_active_index;
+    fwu_md_rss->md.previous_active_index = index;
+
+    index = fwu_md_ap->md.active_index;
+    fwu_md_ap->md.active_index = fwu_md_ap->md.previous_active_index;
+    fwu_md_ap->md.previous_active_index = index;
+
+    priv_metadata->boot_attempted = 0;
+
+    ret = private_metadata_write(priv_metadata);
+    if (ret) {
+        return ret;
+    }
+    fwu_md_rss->md.crc_32 = crc32((uint8_t *)&fwu_md_rss->md.version,
+                              sizeof(struct fwu_metadata_rss) - sizeof(uint32_t));
+
+    fwu_md_ap->md.crc_32 = crc32((uint8_t *)&fwu_md_ap->md.version,
+                              sizeof(struct fwu_metadata_ap) - sizeof(uint32_t));
+
+    ret = metadata_write(FWU_RSS_FLASH_DEV, &fwu_md_rss->md);
+    if (ret) {
+        return ret;
+    }
+
+    ret = metadata_write(FWU_AP_FLASH_DEV, &fwu_md_ap->md);
+    if (ret) {
+        return ret;
+    }
+
+    FWU_LOG_MSG("%s: in regular state by choosing previous active bank\n\r",
+                 __func__);
+
+    FWU_LOG_MSG("%s: exit: ret = %d\n\r", __func__, ret);
+    return ret;
+
+}
+
+static enum fwu_agent_error_t update_nv_counters(
+                        struct fwu_private_metadata* priv_metadata)
+{
+    enum tfm_plat_err_t err;
+    uint32_t security_cnt;
+    enum tfm_nv_counter_t tfm_nv_counter_i;
+
+    for (uint8_t i = 0; i <= FWU_MAX_NV_COUNTER_INDEX; i++) {
+        switch (i) {
+        case FWU_BL2_NV_COUNTER:
+            tfm_nv_counter_i = PLAT_NV_COUNTER_BL1_0;
+            break;
+        case FWU_TFM_NV_COUNTER:
+            tfm_nv_counter_i = PLAT_NV_COUNTER_BL2_1;
+            break;
+        case FWU_FIP_NV_COUNTER:
+            tfm_nv_counter_i = PLAT_NV_COUNTER_BL2_2;
+            break;
+        case FWU_SI_CL2_NV_COUNTER:
+            tfm_nv_counter_i = PLAT_NV_COUNTER_BL2_3;
+            break;
+        case FWU_SI_CL1_NV_COUNTER:
+            tfm_nv_counter_i = PLAT_NV_COUNTER_BL2_4;
+            break;
+        case FWU_SI_CL0_NV_COUNTER:
+            tfm_nv_counter_i = PLAT_NV_COUNTER_BL2_5;
+            break;
+        case FWU_LCP_NV_COUNTER:
+            tfm_nv_counter_i = PLAT_NV_COUNTER_BL2_6;
+            break;
+        case FWU_SCP_NV_COUNTER:
+            tfm_nv_counter_i = PLAT_NV_COUNTER_BL2_7;
+            break;
+        default:
+            FWU_ASSERT(0);
+            break;
+        }
+
+        err = tfm_plat_read_nv_counter(tfm_nv_counter_i,
+                        sizeof(security_cnt), (uint8_t *)&security_cnt);
+        if (err != TFM_PLAT_ERR_SUCCESS) {
+            return FWU_AGENT_ERROR;
+        }
+
+        if (priv_metadata->nv_counter[i] < security_cnt) {
+            return FWU_AGENT_ERROR;
+        } else if (priv_metadata->nv_counter[i] > security_cnt) {
+            FWU_LOG_MSG("%s: updaing index = %u nv counter = %u->%u\n\r",
+                        __func__, i, security_cnt,
+                        priv_metadata->nv_counter[FWU_BL2_NV_COUNTER]);
+            err = tfm_plat_set_nv_counter(tfm_nv_counter_i,
+                                    priv_metadata->nv_counter[FWU_BL2_NV_COUNTER]);
+            if (err != TFM_PLAT_ERR_SUCCESS) {
+                return FWU_AGENT_ERROR;
+            }
+        }
+    }
+
+    FWU_LOG_MSG("%s: exit\n\r", __func__);
+    return FWU_AGENT_SUCCESS;
+}
+
+static void disable_host_ack_timer(void)
+{
+    FWU_LOG_MSG("%s: timer to reset is disabled\n\r", __func__);
+    SysTick->CTRL &= (~SysTick_CTRL_ENABLE_Msk);
+}
+
+enum fwu_agent_error_t fwu_host_ack(void)
+{
+    enum fwu_agent_error_t ret;
+    struct fwu_private_metadata priv_metadata;
+    enum fwu_agent_state_t current_state;
+
+    FWU_LOG_MSG("%s: enter\n\r", __func__);
+
+    if (!is_initialized_rss || !is_initialized_ap) {
+        return FWU_AGENT_ERROR;
+    }
+
+    if (metadata_read(FWU_RSS_FLASH_DEV, &fwu_md_rss.md)) {
+        FWU_ASSERT(0);
+    }
+
+    if (metadata_read(FWU_AP_FLASH_DEV, &fwu_md_ap.md)) {
+        FWU_ASSERT(0);
+    }
+
+    if (private_metadata_read(&priv_metadata)) {
+        ret = FWU_AGENT_ERROR;
+        goto out;
+    }
+
+    current_state = get_fwu_agent_state(&fwu_md_rss, &priv_metadata);
+    if (current_state == FWU_AGENT_STATE_REGULAR) {
+
+        /* Nothing to be done. */
+        ret = FWU_AGENT_SUCCESS;
+
+        fmp_set_image_info(&full_capsule_image_guid,
+                priv_metadata.fmp_version,
+                priv_metadata.fmp_last_attempt_version,
+                priv_metadata.fmp_last_attempt_status);
+
+        goto out;
+
+    } else if (current_state != FWU_AGENT_STATE_TRIAL) {
+        FWU_ASSERT(0);
+    }
+
+    if (fwu_md_rss.md.active_index != priv_metadata.boot_index) {
+
+        /* Firmware update failed, revert back to previous bank. */
+        priv_metadata.fmp_last_attempt_version =
+         fwu_md_rss.img_entry[IMAGE_0].img_props[fwu_md_rss.md.active_index].version;
+
+        priv_metadata.fmp_last_attempt_status = LAST_ATTEMPT_STATUS_ERROR_UNSUCCESSFUL;
+
+        ret = fwu_select_previous(&fwu_md_rss, &fwu_md_ap, &priv_metadata);
+
+    } else {
+
+        /* Firmware update successful. */
+        priv_metadata.fmp_version =
+         fwu_md_rss.img_entry[IMAGE_0].img_props[fwu_md_rss.md.active_index].version;
+        priv_metadata.fmp_last_attempt_version =
+         fwu_md_rss.img_entry[IMAGE_0].img_props[fwu_md_rss.md.active_index].version;
+
+        priv_metadata.fmp_last_attempt_status = LAST_ATTEMPT_STATUS_SUCCESS;
+
+        ret = fwu_accept_image(&full_capsule_image_guid, &fwu_md_rss, &fwu_md_ap,
+                                &priv_metadata);
+        if (!ret) {
+            ret = update_nv_counters(&priv_metadata);
+        }
+    }
+
+    if (ret == FWU_AGENT_SUCCESS) {
+        disable_host_ack_timer();
+        fmp_set_image_info(&full_capsule_image_guid,
+                priv_metadata.fmp_version,
+                priv_metadata.fmp_last_attempt_version,
+                priv_metadata.fmp_last_attempt_status);
+    }
+
+out:
+
+    FWU_LOG_MSG("%s: exit: ret = %d\n\r", __func__, ret);
+    return ret;
+}
+
+static int systic_counter = 0;
+
+void SysTick_Handler(void)
+{
+    systic_counter++;
+    if (systic_counter % 10 == 0) {
+        SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;
+        stdio_output_string("*", 1);
+        SysTick->CTRL |= SysTick_CTRL_ENABLE_Msk;
+    }
+    if (systic_counter == HOST_ACK_TIMEOUT_SEC) {
+        SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;
+        stdio_output_string("timer expired!\n\r",
+                           sizeof("timer expired!\n\r"));
+        tfm_hal_system_reset();
+    }
+}
+
+/* When in trial state, start the timer for host to respond.
+ * Diable timer when host responds back either by calling
+ * fwu_accept_image or fwu_select_previous.
+ * Otherwise, resets the system.
+ */
+void host_acknowledgement_timer_to_reset(void)
+{
+    struct fwu_private_metadata priv_metadata;
+    enum fwu_agent_state_t current_state;
+
+    FWU_LOG_MSG("%s: enter\n\r", __func__);
+
+    if (!is_initialized_rss || !is_initialized_ap) {
+        FWU_ASSERT(0);
+    }
+
+    if (private_metadata_read(&priv_metadata)) {
+        FWU_ASSERT(0);
+    }
+
+    if (metadata_read(FWU_RSS_FLASH_DEV, &fwu_md_rss.md)) {
+        FWU_ASSERT(0);
+    }
+
+    current_state = get_fwu_agent_state(&fwu_md_rss, &priv_metadata);
+
+    if (current_state == FWU_AGENT_STATE_TRIAL) {
+        FWU_LOG_MSG("%s: in trial state, starting host ack timer\n\r",
+                        __func__);
+        systic_counter = 0;
+        if (SysTick_Config(SysTick_LOAD_RELOAD_Msk)) {
+            FWU_LOG_MSG("%s: timer init failed\n\r", __func__);
+            FWU_ASSERT(0);
+        } else {
+            FWU_LOG_MSG("%s: timer started: seconds to expire : %u\n\r",
+                        __func__, HOST_ACK_TIMEOUT_SEC);
+        }
+    }
+
+    FWU_LOG_MSG("%s: exit\n\r", __func__);
+    return;
+}
diff --git a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.h b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.h
index 4724e3217..87d5f4077 100644
--- a/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.h
+++ b/platform/ext/target/arm/rss/kronos/fw_update_agent/fwu_agent.h
@@ -154,10 +154,39 @@ enum fwu_agent_error_t metadata_write(enum FWU_METADATA_FLASH_DEV DEV_TYPE,
                                       struct metadata *p_metadata);
 enum fwu_agent_error_t fwu_metadata_init(enum FWU_METADATA_FLASH_DEV DEV_TYPE,
                                          int *is_initialized);
+/* Host to RSS:
+ * Host responds with this api to acknowledge its successful boot.
+ */
+enum fwu_agent_error_t fwu_host_ack(void);
 void bl1_get_active_bl2_image(uint32_t *bank_offset);
 uint8_t bl2_get_boot_bank(void);
 enum fwu_agent_state_t get_fwu_agent_state(
         struct fwu_metadata_rss *fwu_md_rss_ptr,
         struct fwu_private_metadata *priv_metadata_ptr);

+/* When in trial state, start the timer for host to respond.
+ * Diable timer when host responds back either by calling
+ * fwu_accept_image or fwu_select_previous.
+ * Otherwise, resets the system.
+ */
+void host_acknowledgement_timer_to_reset(void);
+
+enum fwu_nv_counter_index_t {
+    FWU_BL2_NV_COUNTER = 0,
+    FWU_TFM_NV_COUNTER,
+    FWU_FIP_NV_COUNTER,
+    FWU_SI_CL2_NV_COUNTER,
+    FWU_SI_CL1_NV_COUNTER,
+    FWU_SI_CL0_NV_COUNTER,
+    FWU_LCP_NV_COUNTER,
+    FWU_SCP_NV_COUNTER,
+    FWU_MAX_NV_COUNTER_INDEX = FWU_SCP_NV_COUNTER
+};
+
+/* Stage nv counter into private metadata section of the flash.
+ * Staged nv counters are written to the otp when firmware update is successful.
+ * The function assumes that the api is called in the boot loading stage.
+ */
+enum fwu_agent_error_t fwu_stage_nv_counter(enum fwu_nv_counter_index_t index,
+        uint32_t img_security_cnt);
 #endif /* FWU_AGENT_H */
--
2.25.1

